### PR TITLE
Create reusable services

### PR DESCRIPTION
**Pull request by hugobho**
Refactor 'create' command directory to contain both 'application' and 'service' directories
Creates reusable services next to app directory

#75

### CODE REVIEW

**Comment by hugobho** [comment]
@kevgo

File: features/abbreviations.feature [hunk 1]
```diff
@@ -12,13 +12,21 @@ Feature: executing abbreviated Exosphere commands
...
-    Then the full command "exo create" is executed
+  Scenario Outline: executing "exo create application" with abbreviated commands
+    When executing the abbreviated command <CREATE-APP-COMMAND> in the terminal
+    Then the full command "exo create application" is executed
   Examples:
-    | CREATE-COMMAND |
-    | exo c          |
-    | exo creat      |
+    | CREATE-APP-COMMAND |
+    | exo c a            |
```
Comment (kevgo) [inline features/abbreviations.feature]: This is totally awesome!

File: features/create_service.feature [hunk 2]
```diff
@@ -0,0 +1,49 @@
+Feature: create a reusable service
+
+  As an application developer
+  I want to be able to scaffold a reusable service
+  So that I can create business logic that can be used in several apps.
+
+  Rules:
+  - call "exo create servce [<name>] [<description>] [<template>] [<namespace>]" to scaffold a reusable service
```
Comment (kevgo) [inline features/create_service.feature]: typo in "servce"

Comment (kevgo) [inline features/create_service.feature]: I think it's better to move the description to the end. It's the most optional of all the arguments.

File: features/create_service.feature [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      """
+      name: empty app
+      description: Empty test application
+      version: 1.0.0
+
+      services:
+        users-service:
+          location: ../users-service
+      """
+    And the file "users-service/service.yml" exists next to the application directory with content:
```
Comment (kevgo) [inline features/create_service.feature]: And the file "../users-service/service.yml" exists with the content:

File: features/create_service.feature [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      """
+      name: users-service
+      description: manage_users
+
+      setup: pnpm install
```
Comment (kevgo) [inline features/create_service.feature]: This should be "npm install"

File: features/steps/steps.ls [hunk 5]
```diff
@@ -196,6 +203,10 @@ module.exports = ->
     expect(fs.exists-sync path.join(app-dir, file-path)).to.be.true
 
 
+  @Then /^the file "([^"]*)" exists next to the application directory with content:$/ (file-path, expected-content, done) ->
```
Comment (kevgo) [inline features/steps/steps.ls]: I think this step can go away

File: features/steps/steps.ls [hunk 6]
```diff
@@ -213,11 +224,12 @@ module.exports = ->
...
     expected-text = switch command
-      | 'exo run'    => 'exorun'
-      | 'exo test'   => 'exo-test'
-      | 'exo setup'  => 'exo-setup'
-      | 'exo create' => 'We are about to create a new Exosphere application!'
-      | 'exo add'    => 'We are about to add a new Exosphere service to the application!'
+      | 'exo run'                => 'exorun'
+      | 'exo test'               => 'exo-test'
+      | 'exo setup'              => 'exo-setup'
+      | 'exo create application' => 'We are about to create a new Exosphere application!'
```
Comment (kevgo) [inline features/steps/steps.ls]: Please remove the `!` from these strings. Punctuation can change quite frequently, and doing so shouldn't break these tests. We really only care that they output matching text here, not the correct punctuation.

File: features/user_error.feature [hunk 7]
```diff
@@ -13,3 +13,11 @@ Feature: wrong command
   Scenario: the user enters an unknown command
     When running "exo zonk" in the terminal
     Then it prints "Error: unknown command 'zonk'" in the terminal
+
+  Scenario: the user does not specify which "exo create" command
+    When running "exo create" in the terminal
+    Then it prints "Error: missing command" in the terminal
```
Comment (kevgo) [inline features/user_error.feature]: It should ask the user for the missing data in this case

File: features/user_error.feature [hunk 8]
```diff
@@ -13,3 +13,11 @@ Feature: wrong command
...
+
+  Scenario: the user enters unknown "exo create" command
+    When running "exo create zonk" in the terminal
+    Then it prints "Error: missing command" in the terminal
```
Comment (kevgo) [inline features/user_error.feature]: It should display a meaningful error message like "cannot create 'zonk'"

File: src/commands/create/cli.ls [hunk 9]
```diff
@@ -1,34 +1,36 @@
...
-inquirer.prompt(questions).then (answers) ->
-  src-path = path.join __dirname, '..', '..', '..', 'templates', 'create-app'
-  target-path = answers['app-name']
+function unknown-command command
+  console.log red "Error: unknown command '#{command-name}'\n"
+  print-usage!
+
+
+function print-usage
+  console.log 'Usage: exo create <command> [options]\n'
```
Comment (kevgo) [inline src/commands/create/cli.ls]: Isn't the usage more like:

```
exo create [<entity>] [<name>] [<description>] [<template>] [<namespace>]
```

File: src/commands/create/cli.ls [hunk 10]
```diff
@@ -1,34 +1,36 @@
...
+  console.log 'Available commands are:'
+  for command in command-names!
+    console.log "* #{command}"
```
Comment (kevgo) [inline src/commands/create/cli.ls]: We don't need to list all the commands here. The user has already typed `exo create`. We should list all the options that can follow there, i.e. "service" and "application".

File: src/commands/create/commands/application/cli.ls [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+  * type: 'input'
+    name: 'app-version'
+    message: 'Initial version:'
+    default: '0.0.1'
+
+inquirer.prompt(questions).then (answers) ->
+  src-path = path.join __dirname, '..' '..' '..' '..' '..' 'templates' 'create-app'
+  console.log("i am here")
+  console.log(src-path)
```
Comment (kevgo) [inline src/commands/create/commands/application/cli.ls]: delete these two lines

File: src/commands/create/commands/application/cli.ls [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  target-path = answers['app-name']
+  console.log!
```
Comment (kevgo) [inline src/commands/create/commands/application/cli.ls]: delete

File: src/commands/create/commands/service/cli.ls [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+{data, questions} = parse-command-line process.argv
+inquirer.prompt(questions).then (answers) ->
+  data := merge data, answers
+  src-path = path.join __dirname, '..' '..' '..' '..' '..' 'templates' 'add-service' data.template-name
+  target-path = path.join process.cwd!, '..' data.service-name
+  try
+    app-config = yaml.safe-load fs.read-file-sync('application.yml', 'utf8')
+  catch error
+    console.log e
```
Comment (kevgo) [inline src/commands/create/commands/service/cli.ls]: delete

File: src/commands/create/commands/service/cli.ls [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    throw e
```
Comment (kevgo) [inline src/commands/create/commands/service/cli.ls]: End the app with a meaningful error message here, not a stacktrace

File: src/commands/create/commands/service/cli.ls [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+      name: 'serviceName'
+      filter: (input) -> input.trim!
+      validate: (input) -> input.length > 0
+
+  if description
+    data.description = description
+  else
+    questions.push do
+      message: 'Description:'
+      type: 'intput'
```
Comment (kevgo) [inline src/commands/create/commands/service/cli.ls]: typo in "intput"

File: features/create_service.feature (hunk 4 above)
Comment (hugobho) [inline features/create_service.feature]: All the templates under add-service have their setup command set to `pnpm install` in their service.yml files. Should I change all of them?

File: src/commands/create/cli.ls (hunk 9 above)
Comment (hugobho) [inline src/commands/create/cli.ls]: Yes, but there is also the option for a command line prompt, like in 'exo add'. If we display a suggested usage like that here should we also do one for exo add? 

```
exo add [<entity>] [<name>] [<description>] [<template>] [<namespace>]
```

File: src/commands/create/commands/application/cli.ls (hunk 12 above)
Comment (hugobho) [inline src/commands/create/commands/application/cli.ls]: If you're talking about line 32 `console.log!`, that exists in the master version of the SDK. Should I still delete it?

File: features/create_service.feature (hunk 4 above)
Comment (kevgo) [inline features/create_service.feature]: Oh! Please leave it consistent with pnpm for now.

File: src/commands/create/cli.ls (hunk 9 above)
Comment (kevgo) [inline src/commands/create/cli.ls]: Yes, good idea.

File: src/commands/create/commands/application/cli.ls (hunk 12 above)
Comment (kevgo) [inline src/commands/create/commands/application/cli.ls]: Hmm, it might be there to make the command-line more pretty. Let's leave it in then.

**Comment by hugobho** [comment]
@kevgo

**Comment by kevgo** [comment]
🚢

**Pull request closed without merging**